File: src/services/history.ts
// http_client/src/services/history.ts
//
// history commands. The boundary crossing lives in services/invoke.ts
// (CLAUDE.md section 11, rule 3).
import { call } from "@/services/invoke";
import type { HistoryEntry, RecordHistoryInput } from "@/types/history";
import type { ApiError } from "@/types/http";
import type { Result } from "@/types/result";

export function listHistory(): Promise<Result<HistoryEntry[], ApiError>> {
  return call<HistoryEntry[]>("list_history");
}

/** Returns the stored entry, id and timestamp assigned by storage. */
export function recordHistory(entry: RecordHistoryInput): Promise<Result<HistoryEntry, ApiError>> {
  return call<HistoryEntry>("record_history", { entry });
}

export function deleteHistoryEntry(id: string): Promise<Result<void, ApiError>> {
  return call<void>("delete_history_entry", { id });
}

export function clearHistory(): Promise<Result<void, ApiError>> {
  return call<void>("clear_history");
}

File: src-tauri/src/commands/history.rs
// http_client/src-tauri/src/commands/history.rs
use tauri::State;

use crate::commands::dto::{HistoryEntryDto, RecordHistoryInput};
use crate::commands::error::ApiError;
use crate::domain::models::NewHistoryEntry;
use crate::AppState;

macro_rules! blocking {
    ($state:expr, $service:ident, $body:expr) => {{
        let $service = $state.history.clone();
        tauri::async_runtime::spawn_blocking(move || $body)
            .await
            .map_err(|_| ApiError::internal("storage task failed to complete"))?
    }};
}

#[tauri::command]
pub async fn list_history(state: State<'_, AppState>) -> Result<Vec<HistoryEntryDto>, ApiError> {
    let entries = blocking!(state, service, service.list())?;
    Ok(entries.into_iter().map(HistoryEntryDto::from).collect())
}

/// Called after a send settles, success or failure. Returns the stored entry
/// so the panel can prepend it without re-reading the whole list.
#[tauri::command]
pub async fn record_history(
    state: State<'_, AppState>,
    entry: RecordHistoryInput,
) -> Result<HistoryEntryDto, ApiError> {
    let new_entry = NewHistoryEntry::try_from(entry)?;
    let stored = blocking!(state, service, service.record(&new_entry))?;
    Ok(stored.into())
}

#[tauri::command]
pub async fn delete_history_entry(state: State<'_, AppState>, id: String) -> Result<(), ApiError> {
    blocking!(state, service, service.delete(&id))?;
    Ok(())
}

#[tauri::command]
pub async fn clear_history(state: State<'_, AppState>) -> Result<(), ApiError> {
    blocking!(state, service, service.clear())?;
    Ok(())
}
